src/FastBroadcast.jl

```diff
@@ -3,95 +3,167 @@ module FastBroadcast
-        ii, bc, :bc,
+        ii, bc, :bc, ax, :ax
        )
     push!(loopbody_lin.args, :i)
     append!(loopbody_car.args, ii)
 
+    loop_quote = if !(bcc.maybelinear && (indexstyle === IndexLinear))
+        :(@inbounds Base.Cartesian.@nloops $N i dst begin
+            $loopbody_car
+        end)
+    elseif bcc.maybeivdep
+        :(@inbounds @simd ivdep for i in $eachindex(dst)
+            $loopbody_lin
+        end)
+    else
+        :(@inbounds @simd for i in $eachindex(dst)
+            $loopbody_lin
+        end)
+    end
     quote
         $(Expr(:meta,:inline))
+        isfast = true
+        (Base.Cartesian.@ntuple $N dstaxis) = dstaxes
         $(bcc.loopheader)
-        if $islinear($axes(dst), $(bcc.arrays...))
-            if $safeivdep(dst, $(bcc.arrays...))
-                @inbounds @simd ivdep for i in $eachindex(dst)
-                    $loopbody_lin
-                end
-            else
-                @inbounds @simd for i in $eachindex(dst)
-                    $loopbody_lin
-                end
-            end
+        if isfast
+            $loop_quote
         else
-            @inbounds Base.Cartesian.@nloops $(ndims(dst)) i dst begin
-                $loopbody_car
-            end
+            slow_materialize!(dst, bc)
         end
         dst
     end
 end
-@inline fast_materialize(bc::Broadcasted) = fast_materialize!(similar(bc, Base.Broadcast.combine_eltypes(bc.f, bc.args)), bc)
+@inline _get_axes(x) = axes(x)
+@inline _get_axes(bc::Broadcasted) = map(_get_axes, bc.args)
+@inline __index_style(_) = Val{false}()
+@inline __index_style(::IndexLinear) = Val{true}()
+
+# Two arg `_index_style` checks, either stopping with `Val{false}()`, or continues
+@inline _index_style(_, __) = IndexCartesian()
+@inline _index_style(::IndexLinear, x) = _index_style(x)
+@inline _index_style(::IndexLinear, x::Tuple{}) = IndexLinear()
+@inline _index_style(::IndexLinear, x::Tuple{T}) where {T} = _index_style(first(x))
+@inline _index_style(::IndexLinear, x::Tuple{T,S,Vararg}) where {T,S} = _index_style(_index_style(first(x)), Base.tail(x))
+
+@inline _index_style(x) = IndexStyle(typeof(x)) # require `IndexStyle` to be defined
+@inline _index_style(x::Tuple) = IndexLinear()
+@inline _index_style(x::Number) = IndexLinear()
+@inline _index_style(x::Ref) = IndexLinear()
+@inline _index_style(x::AbstractArray) = IndexStyle(x)
+
+@inline _index_style(bc::Broadcasted) = _index_style(_index_style(first(bc.args)), Base.tail(bc.args))
 
-@inline islinear(dst, src) = dst == axes(src)
-@inline islinear(dst, src, srcs::Vararg{AbstractArray,K}) where {K} = dst == axes(src) && islinear(dst, srcs...)
 
-@inline safeivdep(::Array{T}) where {T<:Union{Bool,Base.HWReal}} = true
-@inline safeivdep(::Array{T}, arg1, args::Vararg{Array,K}) where {K,T<:Union{Bool,Base.HWReal}} = safeivdep(arg1, args...)
-@inline safeivdep(args::Vararg{Any,K}) where {K} = false
+@noinline slow_materialize!(dest, bc) = Base.Broadcast.materialize!(dest, bc)
```

Comment:
Will this introduce allocation?

Reply:
I honestly thought it would, but this is what I'm getting on Julia master:
```julia
julia> alloctest_base!(B,A,y,x) = @. B = A * y' + x
alloctest_base! (generic function with 1 method)

julia> alloctest_fastbc!(B,A,y,x) = @.. B = A * y' + x
alloctest_fastbc! (generic function with 1 method)

julia> A = rand(32,32); x = rand(32); B = similar(A);

julia> @benchmark alloctest_base!($B,$A,$x,$x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     251.076 ns (0.00% GC)
  median time:      251.442 ns (0.00% GC)
  mean time:        251.751 ns (0.00% GC)
  maximum time:     330.314 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     369

julia> @benchmark alloctest_fastbc!($B,$A,$x,$x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     196.885 ns (0.00% GC)
  median time:      199.250 ns (0.00% GC)
  mean time:        198.886 ns (0.00% GC)
  maximum time:     252.907 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     681

julia> Ashim = A[1:1,:];

julia> @benchmark alloctest_base!($B,$Ashim,$x,$x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     235.559 ns (0.00% GC)
  median time:      236.119 ns (0.00% GC)
  mean time:        236.349 ns (0.00% GC)
  maximum time:     306.639 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     429

julia> @benchmark alloctest_fastbc!($B,$Ashim,$x,$x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     221.409 ns (0.00% GC)
  median time:      238.326 ns (0.00% GC)
  mean time:        238.428 ns (0.00% GC)
  maximum time:     395.665 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     492
```

Reply:
But we should consider `@inline`-ing it instead. We do a lot of size checks. If inlined, LLVM can presumably use some of those results to skip some checks in the fallback.